src/main/scala/com/databricks/spark/avro/DefaultSource.scala

```diff
@@ -168,7 +193,16 @@ private[avro] class DefaultSource extends FileFormat with DataSourceRegister {
           // Used to convert `Row`s containing data columns into `InternalRow`s.
           private val encoderForDataColumns = RowEncoder(requiredSchema)
 
-          override def hasNext: Boolean = reader.hasNext
+          private[this] var completed = false
+
+          override def hasNext: Boolean = {
+            val r = reader.hasNext
```

Comment:
The task completion callback runs after the task has completed (whether the task has succeeded or failed), so we can assume that this code will no longer be executing when that happens.

You're right that it's probably a good idea to check `completed` at the start of `hasNext` just to be completely defensive here. Let me go ahead and make that change now.